File: src/main/java/edu/practice/datastructures/PriorityQueue.java
package edu.practice.datastructures;

public class PriorityQueue<E extends Comparable<E>> {

    private final Heap<E> heap;

    public PriorityQueue() {
        this(true);
    }

    public PriorityQueue(boolean isMinHeap) {
        this.heap = new Heap<>(isMinHeap);
    }

    public boolean offer(E data) {
        if (data == null) {
            return false;
        }

        heap.insert(data);
        return true;
    }

    public E poll() {
        if (isEmpty()) {
            return null;
        }

        return heap.remove();
    }

    public E peek() {
        if (isEmpty()) {
            return null;
        }

        return heap.peek();
    }

    public int size() {
        return heap.size();
    }

    public boolean isEmpty() {
        return heap.isEmpty();
    }
}

File: src/main/kotlin/datastructures/PriorityQueue.kt
package edu.practice.datastructures

class PriorityQueue<E: Comparable<E>>(isMinHeap: Boolean = true) {

    // In previous Java implementations, `offer` included a null check and returned `false` for null inputs.
    // Kotlin's null safety and type checks make this unnecessary.
    // As a result, `peek` no longer returns `null` as in the Java implementation;
    // instead, it throws an error originating from the Heap Kotlin class.

    private val heap: Heap<E> = Heap(isMinHeap = isMinHeap)

    fun offer(data: E?): Boolean {
        if (data == null) {
            return false
        }
        heap.insert(data)
        return true
    }

    fun poll(): E {
        return heap.remove()
    }

    fun peek(): E {
        return heap.peek()
    }

    fun size(): Int {
        return heap.size()
    }

    fun isEmpty(): Boolean {
        return heap.isEmpty()
    }
}